### ISSUE TITLE
False positives in DoNotUseInsecureDtdProcessing

### ISSUE THREAD

**Issue by JarLob** [dotnet/roslyn-analyzers#1150]
#### Analyzer package

Microsoft.NetFramework.Analyzers

#### Analyzer

DoNotUseInsecureDtdProcessing

#### Repro steps

a) var xmlDoc = new XmlDocument();
b) xmlDoc.Load (PathToXmlFile);
c) var reader = new XmlTextReader (PathToXmlFile);

Build with .NET 4.5.2 and higher

#### Expected behavior

No warnings

#### Actual behavior

Warnings like:
a) Warning	CA3075	An XmlDocument instance is created without setting its XmlResolver property to a secure value.
b) Warning	CA3075	Unsafe overload of 'Load' method.
c) Warning	CA3075	XmlTextReader instance created with insecure default settings.

#### Notes
@qinxgit There are comments in the analyzer like "for .NET framework >= 4.5.2, the default value for XmlResolver property is null", dotnet version is checked and IsSecureResolver is set to true if >= 4.5.2. However in XmlTextReader case for example it checks for
```
                    if (!(env.IsXmlResolverSet | env.IsSecureResolver) ||
                        !(env.IsDtdProcessingSet | env.IsDtdProcessingDisabled))
                    {
```
Do I miss something? Setting resolver to null by default effectively prevents XXE.

BTW, this difference between .NET 4.5.1 and .NET 4.5.2 is not covered by unit tests. There are some tests like
```
using System;
using System.Reflection;               
using System.Xml;   

[assembly: global::System.Runtime.Versioning.TargetFrameworkAttribute("".NETFramework,Version=v4.5.2"", FrameworkDisplayName = "".NET Framework 4.5.2"")]

namespace TestNamespace
```
but the [assembly] attribute has no effect at GetDotNetFrameworkVersion function the analyzer uses to detect .NET version it is supposed to analyze. However it is unit tests' issue (because of the way they reference mscorlib.dll), version detection in IDE works fine.
Unit tests pass because unit tests that verify there is no warning under 4.5.2 lack their counter parts to verify there is a warning under 4.5.1.

Regarding insecure overloads. I might agree that it is better to use an overload that requires to pass XmlReader explicitly in .NET >= 4.5.2 case, but it generates a lot of noise.

**Comment by mavasani** [comment]
@ivanbasov Can you please check if this repros after https://github.com/dotnet/roslyn-analyzers/pull/1327?

**Comment by mavasani** [comment]
@dotpaul to triage (fixed/duped/still repros)

**Comment by zhr1130** [comment]
@mavasani @dotpaul do we have any update on this one ? We are getting some security bugs filed against us due to above issues.

**Comment by dotpaul** [comment]
This rule looks at both XmlResolvers and DTD processing being enabled. Cases B and C are by design, since DTD processing is enabled (but maybe we can come up with a clearer messages).

Case A's message is only for the XmlResolver, so should probably not generate any warnings for >= .NET Framework 4.5.2.

@chloduan since you were looking at this rule, think you have time to look at this? 😃

**Comment by zhr1130** [comment]
Any suggestion on how to fix the warning with using of XmlUrlResolver ? We used XmlUrlResolver  in our case: 
`            XslCompiledTransform _xsl = new XslCompiledTransform();
using (XmlTextReader reader = new XmlTextReader(new StringReader(cabXMlData)) )
            {
                resolver.Credentials = System.Net.CredentialCache.DefaultNetworkCredentials;
                reader.XmlResolver = resolver;
                _xsl.Load(_reader_xsl, settings, null);
                _xsl.Transform(reader, xmlCabData);
            }
`

**Comment by zhr1130** [comment]
@dotpaul can I get some help regarding this issue ? The bug is out of SLA :(

**Comment by dotpaul** [comment]
> @dotpaul can I get some help regarding this issue ? The bug is out of SLA :(

Will follow up with you over email for the code you provided.
